Add dependency on the GRID license for NVIDIA k8s device plugin

yeazelm opened the pull request:
**Description of changes:**
The GRID driver requires a license to fully function. This currently is fetched as a best effort but a node can come up without a "Licensed" driver and fail later after some amount of time. This change a new file to confirm the license has been correctly acquired or is not needed. This includes a drop in for nvidia-k8s-device-plugin.service to depend on this file:
```
ExecStartPre=/usr/bin/test -f /etc/drivers/.grid-licensed
```
This will be before the `ExecStart` so that the device plugin will not register GPUs unless the license is not required or properly configured.

**Testing done:**
WIP - I'm updating the testing with the new approach for every use case but the primary use cases (GRID, open-gpu, and proprietary) all provide the correct experience. 

Built this change and confirmed that it does not run on a G6 (which uses open-gpu, not GRID so this isn't required), fails to have the node become ready when `nvidia-gridd` is misconfigured to not start. And the nodes become ready as expected when `nvidia-gridd` starts running normally.

Normal g6f.xlarge
```
bash-5.1# journalctl -u grid-license-check.service
Dec 03 16:50:39 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 16:50:39 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 16:50:39 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 16:50:39 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 16:50:40 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 16:50:41 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: grid-license-check.service: Deactivated successfully.
Dec 03 16:50:41 ip-192-168-52-46.us-west-2.compute.internal systemd[1]: Finished GRID License Check.
bash-5.1# ls -al /etc/drivers/.grid-licensed
-rw-r--r--. 1 root root 0 Dec  3 16:50 /etc/drivers/.grid-licensedt-2.compute.internal systemd[1]: Finished GRID License Check.
```

Broken nvidia-gridd results in the node becoming ready but in degraded systemd state and no GPUs offered:
```
bash-5.1# systemctl status
● ip-192-168-71-6.us-west-2.compute.internal
    State: degraded
    Units: 461 loaded (incl. loaded aliases)
     Jobs: 3 queued
   Failed: 1 units
    Since: Wed 2025-12-03 21:24:43 UTC; 46min ago
  systemd: 257.9
  Tainted: unmerged-bin
   CGroup: /


bash-5.1# journalctl -u grid-license-check.service
Dec 03 21:24:55 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 21:24:55 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 21:24:55 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 21:24:55 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 21:24:56 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 21:24:56 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 21:24:56 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 21:24:56 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 21:24:57 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 21:24:57 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 21:24:57 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 21:24:57 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 21:24:59 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 21:25:00 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 21:25:00 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 21:25:00 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 21:25:03 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 21:25:03 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Main process exited, code=exited, status=1/FAILURE
Dec 03 21:25:03 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
Dec 03 21:25:03 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: Failed to start GRID License Check.
Dec 03 21:25:06 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Start request repeated too quickly.
Dec 03 21:25:06 ip-192-168-71-6.us-west-2.compute.internal systemd[1]: grid-license-check.service: Failed with result 'exit-code'.
.... 
# this continues as long as the instances is up but no license exists
```

Output from journal on a g6.2xlarge which uses the fallback:
```
bash-5.1# journalctl -u grid-license-check.service
Dec 03 16:52:46 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 16:52:46 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: grid-license-check.service: Skipped due to 'exec-condition'.
Dec 03 16:52:46 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Condition check resulted in GRID License Check being skipped.
Dec 03 16:52:48 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Starting GRID License Check...
Dec 03 16:52:48 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: grid-license-check.service: Skipped due to 'exec-condition'.
Dec 03 16:52:48 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Condition check resulted in GRID License Check being skipped.
bash-5.1# journalctl -u open-gpu-license-fallback.service
Dec 03 16:52:47 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Starting Open GPU GRID License Check Fallback...
Dec 03 16:52:48 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Finished Open GPU GRID License Check Fallback.
bash-5.1# journalctl -u nvidia-k8s-device-plugin
Dec 03 16:52:48 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Starting Start NVIDIA kubernetes device plugin...
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal systemd[1]: Started Start NVIDIA kubernetes device plugin.
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.175860    3114 main.go:235] "Starting NVIDIA Device Plugin" version="unknown"
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.175889    3114 main.go:238] Starting FS watcher for /var/lib/kubelet/device-plugins
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.175949    3114 main.go:245] Starting OS watcher.
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.176276    3114 main.go:260] Starting Plugins.
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.176288    3114 main.go:317] Loading configuration.
Dec 03 16:52:49 ip-192-168-84-108.us-west-2.compute.internal nvidia-device-plugin[3114]: I1203 16:52:49.177360    3114 main.go:342] Updating config with default resource matching patterns.
....

# The file exists to let the device plugin start
bash-5.1# ls -al /etc/drivers/.grid-licensed
-rw-r--r--. 1 root root 0 Dec  3 16:52 /etc/drivers/.grid-licensed
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
kmod-6.1-nvidia-r570: Add grid-license-check

Add a unit that checks for the license to be valid for GRID. The NVIDIA
k8s device plugin requires this unit so if the license is not present,
then the node never offers gpu resources. This prevents a situation
where a node could fail to get a license, join the cluster, and then
later have workloads start to fail due to the unlicensed status.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.12-nvidia-r570: Add grid-license-check

Add a unit that checks for the license to be valid for GRID. The NVIDIA
k8s device plugin requires this unit so if the license is not present,
then the node never offers gpu resources. This prevents a situation
where a node could fail to get a license, join the cluster, and then
later have workloads start to fail due to the unlicensed status.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.12-nvidia-r580: Add grid-license-check

Add a unit that checks for the license to be valid for GRID. The NVIDIA
k8s device plugin requires this unit so if the license is not present,
then the node never offers gpu resources. This prevents a situation
where a node could fail to get a license, join the cluster, and then
later have workloads start to fail due to the unlicensed status.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

arnaldo2792 commented on packages/kmod-6.1-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,26 @@
+[Service]
+Type=oneshot
+ExecCondition=/usr/bin/ghostdog match-nvidia-driver grid
+# Otherwise, attempt to load the module.
```
I'm confused, will this line actually load the module or are you just generating output that will be `grep`ed later?

yeazelm replied:
This comment might make more sense above the ExecCondition
```suggestion
# Confirm GRID is required, then attempt to check the license
```

arnaldo2792 commented on packages/kmod-6.12-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,26 @@
+ExecCondition=/usr/bin/ghostdog match-nvidia-driver grid
+# Otherwise, attempt to load the module.
+ExecStart=/usr/bin/nvidia-smi -q
+# Ensure that the stderr file exists. Otherwise, grep fails on an empty file.
```
It's the STDOUT file what you are creating, right?

yeazelm replied:
Yes, this is a forgotten update, I moved to STDOUT but forgot to update the comment.

bcressey commented on packages/kmod-6.1-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,26 @@
+StartLimitBurst=120
+
+[Install]
+RequiredBy=kubelet.service
```
Below it's required by `nvidia-k8s-device-plugin.service` which seems more correct.

yeazelm replied:
Sorry, this was originally `nvidia-k8s-device-plugin.service` and I missed the refactor. In my mind we have several options for what experience we want:
* Don't prevent anything but log it. Sort of the behavior we have today just without logs (not ideal)
* Prevent kubelet from starting if the license isn't valid. This means the node never becomes ready and cannot accept any orchestrated work. (This is what I was intending to enforce in the PR)
* Prevent the device plugin from running, the node will become ready but not advertise GPU resources. It could take some work but not GPU work. (This might be confusing to users so I leaned away from this but was my original approach)
* Attempt to block boot entirely so the node doesn't even reach multi-user. (Seems harsh but would be fine too).

yeazelm replied:
I decided to go with nvidia-k8s-device-plugin.service so the instance comes up and is able to be investigated but won't advertise GPUs.

bcressey commented on packages/kmod-6.1-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,26 @@
+# Otherwise, attempt to load the module.
+ExecStart=/usr/bin/nvidia-smi -q
+# Ensure that the stderr file exists. Otherwise, grep fails on an empty file.
+ExecStart=-/usr/bin/touch /tmp/.nvidia-gridd-license
```
`truncate` may be better than `touch` to reset the file between iterations

(would have to move this before the `nvidia-smi` call)

yeazelm replied:
I like that, I'll give it a shot to see if it still gives me the behavior I want with things rearranged to use `truncate`

bcressey commented on packages/kmod-6.12-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,26 @@
+StartLimitBurst=120
+
+[Install]
+RequiredBy=nvidia-k8s-device-plugin.service
```
Does this requirement cause the k8s device plugin to fail if the license check fails? I'm kind of worried about cluttering up the logs with a lot of failures.

This could possibly be modeled as:
1. a timer unit that runs and creates a marker file when the license check passes
2. a path unit that activates `nvidia-k8s-device-plugin.service`
3. a fallback unit that runs when we don't match the grid driver that also creates the marker
4. a condition in the k8s device plugin that requires the marker to exist

yeazelm replied:
This might be much cleaner in the logs, otherwise the unit is very angry and noisy in the journal when its failing. I'll play with that as a potential alternative to this. FWIW though I haven't seen this fail yet before the next unit runs when we are going to get a license, so it might be a situation where the only time its noisy, is when the node is already in a bad state. Nonetheless, I think making it cleaner is worth it.

yeazelm replied:
This is the new approach.

Commit:
kmod-6.1-nvidia-r580: Add grid-license-check

Add a unit that checks for the license to be valid for GRID. The NVIDIA
k8s device plugin requires this unit so if the license is not present,
then the node never offers gpu resources. This prevents a situation
where a node could fail to get a license, join the cluster, and then
later have workloads start to fail due to the unlicensed status.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

yeazelm commented:
^ I updated the approach to use timers and a marker file. I'm working on all the new logs from testing to confirm the behavior works as described.

bcressey commented on packages/kmod-6.1-nvidia-r570/grid-license-check.timer:
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=GRID License Check Timer
+RefuseManualStart=true
+DefaultDependencies=no
```
Why `DefaultDependencies=no`? 
```suggestion
RefuseManualStart=true
```

We need to be careful that this timer doesn't queue a start job for something ahead of schedule. In this case, it could cause all the dependencies of `grid-license-check.service` to start before we're ready, at least on FIPS variants where we start in `fipscheck.target` rather than `preconfigured.target`.

bcressey commented on packages/kmod-6.1-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,20 @@
+ExecStart=/usr/bin/touch /etc/drivers/.grid-licensed
+ExecStart=/usr/bin/systemctl stop grid-license-check.timer
+StandardOutput=append:/tmp/.nvidia-gridd-license
+
+[Install]
+WantedBy=nvidia-k8s-device-plugin.service
```
Do we need this dependency?

yeazelm replied:
I moved to the drop in

bcressey commented on packages/kmod-6.1-nvidia-r570/tesla-license-fallback.service:
```diff
@@ -0,0 +1,15 @@
+
+[Service]
+Type=oneshot
+ExecCondition=!/usr/bin/ghostdog match-nvidia-driver tesla
```
I'm not clear on what this is meant to be doing.

`ExecCondition` doesn't support negation, and if it did, this would mean "execute if we're not using the tesla driver" which seems like the opposite of what we want.

bcressey commented on packages/kmod-6.1-nvidia-r570/grid-license-check.service:
```diff
@@ -0,0 +1,20 @@
+ExecStart=/usr/bin/nvidia-smi -q
+ExecStart=/usr/bin/grep -q "License Status.*: Licensed" /tmp/.nvidia-gridd-license
+ExecStart=/usr/bin/touch /etc/drivers/.grid-licensed
+ExecStart=/usr/bin/systemctl stop grid-license-check.timer
```
I'd always recommend `--no-block` unless you're very sure this could never possibly deadlock:
```suggestion
ExecStart=/usr/bin/systemctl stop grid-license-check.timer --no-block
```

bcressey reviewed (commented):
For this:
> A change will need to be added to the nvidia-k8s-device-plugin.service to depend on this file

Can you add this as a drop-in so there's no cross-kit dependency?

yeazelm commented:
^ Updated for the comments and added in the drop-in for nvidia-k8s-device-plugin.

bcressey commented on packages/kmod-6.12-nvidia-r580/tesla-license-fallback.service:
```diff
@@ -0,0 +1,14 @@
+Type=oneshot
+ExecCondition=/usr/bin/ghostdog match-nvidia-driver tesla
+ExecStart=/usr/bin/touch /etc/drivers/.grid-licensed
+ExecStart=/usr/bin/systemctl stop grid-license-check.timer
```
```suggestion
ExecStart=/usr/bin/systemctl stop grid-license-check.timer --no-block
```

bcressey commented on packages/kmod-6.1-nvidia-r580/tesla-license-fallback.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Tesla GRID License Check Fallback
+RefuseManualStart=true
+RefuseManualStop=tru
```
```suggestion
RefuseManualStop=true
```

bcressey reviewed (approved):
Two minor fixes, LGTM otherwise.

yeazelm commented:
^ Fixed typos from comments/

arnaldo2792 commented on packages/kmod-6.1-nvidia-r570/grid-license-file-check.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+ExecStartPre=/usr/bin/test -f /etc/drivers/.grid-licensed
```
Why not `ConditionPathExists`?

https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#ConditionPathExists=

yeazelm replied:
`ConditionPathExists` will fail the unit, I want systemd to retry. If the unit happens to get scheduled before we have placed this file, it would just mark it as failed. The `ExecStartPre` matches the rest of the services dependencies. See https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/develop/packages/nvidia-k8s-device-plugin/nvidia-k8s-device-plugin.service for the other pre conditions.
